### ISSUE TITLE
Trivial Support For Resharding (RFC)

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#45766]
**Concept**

Resharding can be implemented by extending `ReplicatedMergeTree` table engine a little, so there is a replica on every cluster node, but every replica is responsible for an automatically selected subset of table partitions (so the notions of "replica" and "shard" are combined; and the notions of "partition key" and "sharding key" are also combined).

For example, you have a cluster of 10 nodes, and you create a Replicated table on all of them, specifying that the new placement algorithm should be used with replication factor 3. You also specify `PARTITION BY (toYYYMM(date), UserID % 16)`, so you have 16 partitions for every month of data.

Whenever a new partition appears, three replicas are selected to process it. Selection of the replicas can be made using the host anti-affinity tag (such as rack or datacenter name), the number of other partitions on the replica, the amount of free space, and on a first-come basis. Every partition is assigned to at least three replicas but can be assigned to a larger number of replicas if the partition is moved.

When a replica is unavailable or lagging for a prolonged time, or when is a significantly unbalanced distribution of data between replicas (for example, after adding new hosts), resharding can be initiated by moving partitions between replicas. It is being done by increasing the number of replicas per partition by adding a new replica to be responsible for partition and optionally removing the old replica.

Every replica is subscribed to the same metadata in Keeper (as for normal Replicated tables) but processes only a subset of partitions to which it is assigned. Every replica also acts like a Distributed table for INSERT, SELECT, and other queries. SELECT will choose a set of replicas having the required partitions with not too high replication lag and request to read a subset of partitions on each of them, so the complete set of data without duplicates will be queried. INSERT will split data into partitions, choose available replicas for every partition and synchronously send data there.

**Out of scope**

Partitioning by dynamic ranges. Scalability to a very large number of nodes.

**Implementation Proposal**

A feature flag to enable the new replication mode.

Make every replica post additional status about itself in Keeper: host anti-affinity tag, the amount of free space, and the number of partitions assigned.

Add a node in Keeper with the list of partitions as children and the list of assigned replicas for every partition.

Support for "filtered replication". A replica will process the same log but only store the data parts belonging to the subset of partitions.

Selecting the replicas when a new partition appears on INSERT by using a multi-transaction in Keeper.

Implementation of distributed INSERT and SELECT queries with respect to the data placement. Note: SELECT queries can avoid synchronous requests to Keeper in favor of asynchronous update of the data placement map if the removal of old replicas will be performed with a delay.

Data rebalancing by "cloning" partitions from other replicas and removing partitions from other replicas.

**Comment by UnamedRus** [comment]
It seems, that we need to have lazy load primary key for that: https://github.com/ClickHouse/ClickHouse/issues/43424

Otherwise, if each replica will hold 10x of current data, PRIMARY KEY size will became an issue.

**Comment by alexey-milovidov** [comment]
@UnamedRus 

> Otherwise, if each replica will hold 10x of current data, PRIMARY KEY size will became an issue.

Each replica will be responsible for a subset of partitions so that the data will be spread across the cluster with the desired replication factor.

**Comment by UnamedRus** [comment]
Bucketing is good idea, and it's actually implemented in such way in clickhouse derived dbms like [ByConity](https://github.com/ByConity/ByConity)

Only difference is: they use new syntax for that and doesn't reuse partition statement.

```
CREATE TABLE [IF NOT EXISTS] [db.]table_name [ON CLUSTER cluster]
(
    name1 [type1] [DEFAULT|MATERIALIZED|ALIAS expr1] [TTL expr1],
    name2 [type2] [DEFAULT|MATERIALIZED|ALIAS expr2] [TTL expr2],
    ...
    INDEX index_name1 expr1 TYPE type1(...) GRANULARITY value1,
    INDEX index_name2 expr2 TYPE type2(...) GRANULARITY value2
) ENGINE = MergeTree()
ORDER BY expr
[PARTITION BY expr]
[CLUSTER BY expr INTO <TOTAL_BUCKET_NUMBER> BUCKETS [SPLIT_NUMBER <SPLIT_NUMBER_VALUE>] [WITH_RANGE] ]
[PRIMARY KEY expr]
[SAMPLE BY expr]
[TTL expr [DELETE|TO DISK 'xxx'|TO VOLUME 'xxx'], ...]
[SETTINGS name=value, ...]
```

One potential problem i see with that bucketing is:

* INSERT fan-out, when single INSERT will create hundreds of parts (if we have total number of buckets ~ 256).

And we "need" to have somewhat big total bucket count if we want to have more equal data distribution or/and smaller data rebalances.

**Comment by helifu** [comment]
It seems [apache kudu](https://kudu.apache.org/docs/index.html#_architectural_overview) is a very good example for storage layer. Maybe we can introduce a new central role (`master`) to manage the cluster.. At the same time, keep the zookeeper for data replication. Bt the way, doris's partitioning rule is a somewhat similar to kudu's.

**Comment by ylw510** [comment]
@alexey-milovidov Hi, Could you share the current status or development progress of this feature? Is it already under development, or has development not started yet?

**Comment by zoomxi** [comment]
Here's a concrete design and phased delivery plan for this RFC, with a working prototype at #102244 covering the full scope below.

## TL;DR
  ```sql
  CREATE TABLE events ON CLUSTER '{cluster}' (...)
  ENGINE = ReplicatedMergeTree('/tables/events', '{replica}')
  PARTITION BY (toYYYMM(date), ID % 16)
  SETTINGS replication_factor = 2;   -- each partition on 2 of N replicas
```
Extend `ReplicatedMergeTree` with a `replication_factor` MergeTree setting. When > 0, each partition is assigned to exactly N replicas via Keeper; SELECT routes to assigned replicas, INSERT forwards blocks to assigned replicas, merges run only on assigned replicas. Unassigned replicas still participate in the shared log but skip partition-scoped work. Rebalancing is a CLONE → SWITCH → DONE state machine coordinated through Keeper.

## Mapping to the original scope
```
  ┌─────────────────────────────────┬────────────────────────────────────────────────────────────────────────────────────────────┐
  │        Original RFC item        │                                       Design answer                                        │
  ├─────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────┤
  │ Feature flag to enable new mode │ replication_factor (MergeTree setting, EXPERIMENTAL, default 0)                            │
  ├─────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────┤
  │ Keeper metadata per replica     │ Reuses existing /replicas/{name}/*; adds /selective/counts for load                        │
  ├─────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────┤
  │ Per-partition assignment        │ /selective/assignments/{partition_id} — sorted replica list, versioned                     │
  ├─────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────┤
  │ Filtered replication            │ ReplicatedMergeTreeQueue::shouldSkipForSelectiveReplication + merge-strategy picker filter │
  ├─────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────┤
  │ Replica selection on INSERT     │ Least-loaded-first + hash rotation; atomic multi-op CAS-create                             │
  ├─────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────┤
  │ Distributed INSERT / SELECT     │ InsertForwarder (sink-level) and Router (read path), depth-limited to prevent loops        │
  ├─────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────┤
  │ Rebalancing by cloning          │ MigrationCoordinator: CLONE → SWITCH → DONE with :cloning suffix in assignment             │
  ├─────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────┤
  │ Async assignment map for SELECT │ Per-replica TTL cache (default 60s), refreshed after pullLogsToQueue                       │
  └─────────────────────────────────┴────────────────────────────────────────────────────────────────────────────────────────────┘
```

## Basic Design
### 1. Keeper layout

```
  /{zookeeper_path}/
  ├── ... existing ReplicatedMergeTree nodes (replicas/, log/, parts/, ...)
  └── selective/                                ← NEW, only created when replication_factor > 0
      ├── config                                ← {"replication_factor": N, "format_version": 1}
      ├── counts                                ← "format version: 1\nreplica_a:120,replica_b:118,..."
      ├── assignments/
      │   ├──partition_id_1                     ← "format version: 1\nreplica_a,replica_c"
      │   ├──partition_id_2                     ← "format version: 1\nreplica_b,replica_c"
      │   └── ...
      ├── migrations/
      │   └── <uuid>                            ← JSON: {state, partition_id, source, target, ...}
      └── rebalance_lock
```
  - config — replication factor + format version. Validated against local MergeTreeSetting::replication_factor at startup.
  - counts — per-replica partition count. Read once during allocation; CAS-updated in the same multi-op as the assignment CAS-create. Avoids a full scan of /assignments on the hot path.                                                                                                                                                           
  - assignments/{partition_id} — the authoritative placement. Readers cache with TTL; writers use CAS.
  - migrations/{uuid} — one node per active migration; JSON-serialized state machine.

### 2. Read path
```
  SELECT at coordinator (distributed_depth = 0)
    ↓
  Router checks partition → replica assignment from local cache
    ↓
  Partitions split into {local, remote_by_replica}
    ↓
  Local partitions: MergeTreeDataSelectExecutor
    ↓
  Remote partitions: ReadFromRemote to assigned replica, tagged with `_partition_id IN (...)`
    ↓
  At remote (distributed_depth > 0): Router verifies its cache still owns those
  partitions; any misplaced ones re-forward (bounded by MAX_FORWARDING_DEPTH = 5)
```
  Routing decisions use a TTL cache (default 60s); stale cache never causes incorrect results because the depth-gated re-routing path handles disagreement.

### 3. Write path
```
  INSERT
    ↓
  Sink receives blocks, splits by partition
    ↓
  [Phase 1] For each partition, `ensurePartitionAssignment` (CAS-create if missing)
    ├─ local partitions → commitPart() with CAS-check on assignment node version
    └─ remote partitions → forward block to one assigned replica (connection pool)
    ↓
  [Phase 2] On CAS mismatch (rare — assignment changed mid-insert):
    re-read assignment, re-route misplaced blocks
```
### 4. Rebalancing
State machine per migration:
```
     startClone (manual: SYSTEM MIGRATE PARTITION, or auto: rebalance plan)
         │
         │   multi-op: create /migrations/<uuid> + append "target:cloning" to assignment
         ▼
     ┌─────────┐  target fetches parts   ┌──────────┐  source removed from   ┌──────┐
     │  CLONE  │ ──────────────────────▶ │  SWITCH  │ ──────────────────────▶│ DONE │
     └─────────┘                         └──────────┘   assignment           └──────┘
         │                                    │
         │ timeout or error                   │ rollback
         ▼                                    ▼
     ┌─────────┐                         (revert :cloning)
     │ FAILED  │
     └─────────┘
```
  - :cloning suffix in the assignment list signals "target is copying, still counts as N+1 until clone completes" — readers tolerate it, writers forward to it too.
  - Coordinator is just the replica that created the migration node; if it dies, recoverOrphanedMigrations takes over via Keeper CAS after migration_coordinator_timeout (default 60s).

## 5. Delivery roadmap

One foundational PR establishes the placement invariant ("partition P's data physically exists only on `assignments[P]`"), then three smaller PRs add operational capabilities on top.

  | PR | Scope | User-visible increment |
  |---|---|---|
  | **1 (foundation)** | Assignment allocation, read routing, INSERT forwarding, queue-skip, merge-strategy filter | Complete selective storage on static clusters: storage, read, and write fan-out all reduce from N to `replication_factor` |
  | **2** | `OPTIMIZE` forwarding |  `OPTIMIZE` targets the right replicas |
  | **3** | Manual migration (`SYSTEM MIGRATE PARTITION`) | Ops can manually rebalance during scale-up / scale-down |
  | **4** | Auto rebalance monitor (`SYSTEM START SELECTIVE REBALANCE`, orphan recovery) | Automatic rebalancing on membership change |

PR 1 is larger than expected as I couldn't find a cleaner seam. This initial PR needs to be relatively complete to maintain consistency. Suggestions on how to split it further are more than welcome.
